Refactor GCEClient: wrap compute.Service in an interface for mocking …

spew opened the pull request:
…GCP compute

This change creates a ComputeService interface which has a runtime
implementation that wraps the compute.Service. This will enable creating
tests that mock GCP Compute Service calls.

**What this PR does / why we need it**:
It is difficult to test changes that interact with GCP without manually running integration tests. This change will enable developers to write unit tests that mock GCP and test that their code correctly handles various scenarios. 

**Release note**:
```release-note
NONE
```

<!-- All reviews default to cc'ing the kube-deploy-reviewers github group. -->
@kubernetes/kube-deploy-reviewers

spew committed:
Refactor GCEClient: wrap compute.Service in an interface for mocking GCP compute

This change creates a ComputeService implementation which has a runtime
implementation that wraps the compute.Service. The MachineActuator is
changed to make use of the ComputeService through a new interface named
GCEClientComputeService. This will enable creating tests that mock GCP Compute
Service calls to control MachineActuator behavior..

jessicaochen reviewed (approved):
lgtm

/ok-to-test

krousey commented on cluster-api/cloud/google/machineactuator.go:
```diff
@@ -59,12 +60,12 @@ type SshCreds struct {
-	kubeadmToken  string
-	sshCreds      SshCreds
-	machineClient client.MachineInterface
+	computeService clients.ComputeService
```
The interface should be defined here. This is the consumer of the functionality and is best aware of what functionality it needs.

krousey commented on cluster-api/cloud/google/clients/computeservice.go:
```diff
@@ -0,0 +1,71 @@
+	service *compute.Service
+}
+
+func NewComputeService(client *http.Client) (ComputeService, error) {
```
You should delete the `ComputeService` interface from this package, change `computeServiceImpl` to `ComputeService`, and return a pointer to it here.

krousey reviewed (changes requested):
Interfaces should really be defined by the consumer, and functions shouldn't return interfaces when they can return just the implementation. There are exceptions, but I don't believe this is one.

Let's assume we have this code here. And we write unit tests that fake out this interface. And these unit tests test functions that consume this interface.

Now let's assume we need to add the ability to set some project metadata for a new piece of code. We have to add the method to the implementation, the interface, and every fake that implements that interface so the functions that took that interface (the ones that didn't change because they don't really care about project metadata) can continue to use the fake.

By moving the interface definition to the consumer, I can have a piece of code that only needs

```go
type MyImageGetter interface {
    ImagesGet(...)
}
```

And as long as any implementation (real or fake) has that method, it can be assigned to it.

krousey commented:
In fact, the Go standard library fell into this trap. When https://golang.org/pkg/compress/flate/ first started out, they returned `io.ReadCloser`. Then they needed to add a `Reset` optimization, but couldn't break the API contract. They had to return an `io.ReadCloser`. Since they couldn't update the return type, they had to do a less than optimal design. They had to add (flate.Resetter)[https://golang.org/pkg/compress/flate/#Resetter] so you could type assert to it, and then call `Reset` as they do in the reset example.

If they had just returned a concrete type that had `Read` and `Close`, users could have still assigned it to an `io.ReadCloser`, and they could have added `Reset` later without breaking the API contract (because they'd still be returning the same struct).

spew commented:
I think i'm coming from the world where creating an interface is expensive, not cheap like Go. Thanks for the justification Kris, I can see how it is much better not have a single 'master' interface but to instead allow consumers to define the interface and then testers of those consumers can implement their own mocks.

Updating as per comments.

spew commented:
I moved the interface into the machineactuator.go file and renamed it to GCEClientComputeService. I pushed a new version of the PR which has an updated commit message to reflect the changes.

jessicaochen reviewed (approved):
lgtm

jessicaochen commented:
@krousey - why does this need migration? It is squarely in the cloud directory.

spew commented:
@jessicaochen we are going with a single repository now.

roberthbailey commented:
@spew - do we still need this change (in the cluster-api repo)? 

As this PR is no longer relevant I'm going to close it.

spew commented:
Already completed in the new repo.
